### ISSUE TITLE
Questions: about the future and some changes

### ISSUE THREAD

**Issue by InExtremaRes** [js-joda/js-joda#405]
Hi! I have some changes I would like to test and see if you are happy with. Some are those from #355 and other small cleanups. But, before starting, I have some questions:

1. Are you still planing to maintain this library for the foreseeable future? I bet you are aware of the new temporal proposal for ES. I think it's a good movement and I happy to read is going forward; I like a lot some aspects of its API, especially since they fill very natural in the ES environment, but I have to say I don't like all of the decisions, for instance I don't like the fact it doesn't provide (yet?) an equivalent for `ZonedDateTime`. Of course since it's just in stage 2 lot of things can change... Anyhow, even if that proposal makes its way through I like this library and its API, I find it very useful and clear (although I must confess it is "too java" IMO :laughing:). I would like to keep working and using this library a lot more.

2. The changes I want to present could be, and very likely will be, suited for a major version. I wouldn't like to send a large PR, instead a prefer to send more, smaller ones that could be discussed independently. I was thinking it might be better to open a new branch so I can open the PRs against that branch instead of `master`; when they're all discussed and hopefully approved, the branch can be merged into `master` What do you think about this?

3. Is the goal of this project to mirror as close as possible the API of the threeten? I ask this since the [README](https://github.com/js-joda/js-joda/blob/master/packages/core/README.md#future-milestones) talks about transform the API to a more "JavasScript" style. Sometimes there are places were the java background of the original API just get in the way, even if a more _javascripty style_ can be used. Of course I understand this is something that should be evaluated case by case, but if your answer is on the line of "no, this should mirror the API" then I'll limit the amount of proposal that change it. When in doubt I will open an issue to ask your opinion, but sometimes is just easier to make a PR and show the code to discuss. (note: In https://github.com/js-joda/js-joda/issues/355#issuecomment-544355045 @pithu already said "in doubt we follow ThreeTen", I know; just asking mostly about those words on the given link to the README and want to know if those are still goals of this project).

**Comment by phueper** [comment]
Hi @InExtremaRes,

I'll try to answer your questions from my point of view, of course @pithu might disagree

Re 1. Yes, we plan on continuing to maintain the library, sadly I currently have very limited time. We are aware of the temporal proposal and have expressed our oppinion in some of the issues, though I have to admit that i haven't been following the proposal very closely lately. Our idea was to see if we can adapt part of the proposal once it is finalized while still keeping the threeten api for js-joda

Re 2. A separate branch might be a good idea for bigger changes, though the rate of change in master is currently pretty slow and you might as well just fork master and post PRs from your fork... But if you feel a branch might make more sense to consolidate multiple PRs, it might be the way to go
Re 2 Yes, the plan and goal is to mirror the threeten api, although we might consider changes that make sense from a js point of view, but I'll have to warn you, there will have to be good arguments for it 😁. That said I think some of the internals could be changed to fit js better, we pretty much tried to port the Java code originally, there might be room for improvement. I think it makes sense to have some pr/code proposals so we can discuss them, I am certainly open to discussion 😊

Regards, Pattrick

**Comment by InExtremaRes** [comment]
Thank you @phueper. I'll wait to @pithu for his opinion.

> Re 2. A separate branch might be a good idea for bigger changes, though the rate of change in master is currently pretty slow and you might as well just fork master and post PRs from your fork... But if you feel a branch might make more sense to consolidate multiple PRs, it might be the way to go

I have a fork, but I would prefer if new branch is opened to send some PRs against to. That way I can send a series of independent changes to discuss, see which of them are merged and send some more. I don't pretend to change the world in PRs nor to have that branch opened for a long time but I'll feel more comfortable that way.

**Comment by InExtremaRes** [comment]
I already have a PR I would like to propose, a little one with changes to names of methods and documentation. Let me know what branch should I open that to.

**Comment by pithu** [comment]
I agree with @phueper but would like to add some things. 

Re 1) IMO the future of js-joda is to be the reference implementation of the ThreeTen API in javascript. I would even think about, to rename the library from `js-joda` to `theeten.js`. Over the past years, we had a lot of requests from the Java community, cross compiler projects, etc and we always promised that we will stick to the ThreeTen API. 

And yes, we want to support this library in the future, but from my side, this will be mainly improvements under the hood, bug fixes and to follow up ThreeTen.

Re 3) I totally agree, that js-joda is to JAVA like and it would be great to make it more javascript like. 

But with the temporal proposal the situation changed. Let's assume that the temporal proposal gets through, is there a place for a third domain-driven immutable date-time API?

Might be, that an "enhanced javascript threeten API” still would make sense, close to threeten and javascript like. But from the experience with the temporal proposal, i guess this would be an extremely time consuming process. Unfortunately i don't see, that i have the time for leading such a process. 

Nevertheless, i always thought, this "enhanced javascript threeten API” should be a thin wrapper on top of ThreeTen. So i don’t see any problem, to start with a project like that. Could be in any repository or as another sub project of js-joda. Assuming somebody takes that challenge.

**Comment by InExtremaRes** [comment]
In the main project I'm working on currently, we benefit a lot from the clear and focused API of ThreeTen (even when none of us have a java background). The temporal proposal is absolutely necessary in my POV but could be a long time until it is mature enough. Besides it lacks some expressive and granular power (as it is now).

To be completely honest, if happens that the proposal is finally accepted and deployed, a library like this would probably be less useful to most people. Most devs just need a way to express a little arithmetic and formatting/parsing, not a full date-time domain model. There is also the fact that most JS libraries are not strongly typed, the temporal proposal could be one of the first going that way and introducing and requiring specific instances of some classes (and I've seen issues discussing a way to weakening the requirements a bit). But for us that strictness is key since at least one of our subdomains requires that clarity of thought and restriction about date-time; we are programming the business logic of a non-trivial domain in TypeScript, not just using JS for presentation (that's the reason I started to notice some issues and reporting/fixing here depending on my time, mainly with discrepancies in the types since the implementation is quite good).

We'll be using this library for much more time and I'm happy to hear the plans are to maintain it. I'll be glad to help a bit according to my own time. For now, in the short term, I'll try to send some changes and improvements trying to match the original ThreeTen API with changes only when I see fits.

BTW, I remember read in other issues you guys talking about "changing the way this library formats dates-times" since it is not that usual in JS. Don't remember where exactly. What do you mean by that? Have any example of your ideas in that regard?

> Nevertheless, i always thought, this "enhanced javascript threeten API” should be a thin wrapper on top of ThreeTen. So i don’t see any problem, to start with a project like that.

Here are you talking about keeping the a port of the ThreeTen as close as possible and, over that, make a wrapper that fits better/closer to JS world?

Also, as I said before, I have already a little PR with some changes (not the one I already opened). Are you @pithu ok with opening a new branch or you prefer I target master all the time?

Thank you both to make the time to answer and discuss.

**Comment by pithu** [comment]
> BTW, I remember read in other issues you guys talking about "changing the way this library formats dates-times" since it is not that usual in JS. Don't remember where exactly. What do you mean by that? Have any example of your ideas in that regard?

Can't remember, can you point to the issue

> Here are you talking about keeping the a port of the ThreeTen as close as possible and, over that, make a wrapper that fits better/closer to JS world?

Yes exactly

> Also, as I said before, I have already a little PR with some changes (not the one I already opened). Are you @pithu ok with opening a new branch or you prefer I target master all the time?

Yes, small junks per PR are helpful, we are always open for improvements. If your target master, it might be easier to merge. i am not a big fan of long running branches, because of many reasons, but it's up to you.

**Comment by InExtremaRes** [comment]
> Yes, small junks per PR are helpful, we are always open for improvements. If you target master, it might be easier to merge. i am not a big fan of long running branches, because of many reasons, but it's up to you.

I'll see how it goes targeting master. If things becomes difficult to me for any reason, I'll then ask you to open a branch :)

> > BTW, I remember read in other issues you guys talking about "changing the way this library formats dates-times" since it is not that usual in JS. Don't remember where exactly. What do you mean by that? Have any example of your ideas in that regard?
> 
> Can't remember, can you point to the issue

This is an example of something like: that https://github.com/js-joda/js-joda/issues/103#issuecomment-255420521 (not exactly what I said, but on the line). I quote from the issue:

> And there are some places in the code, that could be solved much more elegant with a javascript style, like in the DateTimeFormatter.

What did you have in mind at that time? Can you remember?
